swift/ql/lib/codeql/swift/regex/internal/ParseRegex.qll

```diff
@@ -277,35 +277,54 @@ abstract class RegExp extends Expr {
   private predicate isGroupStart(int i) { this.nonEscapedCharAt(i) = "(" and not this.inCharSet(i) }
 
   /**
-   * Holds if a parse mode starts between `start` and `end`.
+   * Holds if the initial part of a parse mode, not containing any
+   * mode characters is between `start` and `end`.
    */
-  private predicate flagGroupStart(int start, int end) {
+  private predicate flagGroupStartNoModes(int start, int end) {
     this.isGroupStart(start) and
     this.getChar(start + 1) = "?" and
     this.getChar(start + 2) in ["i", "x", "s", "m", "w"] and
     end = start + 2
   }
 
   /**
-   * Holds if a parse mode group is between `start` and `end`, and includes the
-   * mode flag `c`. For example the following span, with mode flag `i`:
+   * Holds if `pos` contains a mode character from the
+   * flag group starting at `start`.
+   */
+  private predicate modeCharacter(int start, int pos) {
+    this.flagGroupStartNoModes(start, pos)
+    or
+    this.modeCharacter(start, pos - 1) and
+    this.getChar(pos) in ["i", "x", "s", "m", "w"]
```

Comment:
I couldn't find any official list of parse modes, but Bing Chat says there's also a "u" for Unicode (citing https://developer.apple.com/documentation/swift/regex, which doesn't list *any* flags).

Where did the list in this PR come from?

Reply:
Unfortunately regex parsers vary slightly, and Bing Chat appears to be ignoring this and giving generic advice.

The list comes from the [`NSRegularExpression`](https://developer.apple.com/documentation/foundation/nsregularexpression) doc (search for `(?ismwx-ismwx: ... )`), though I made an assumption that [`Regex`](https://developer.apple.com/documentation/swift/regex) works exactly the same way despite, as you noted, the doc not saying.

I've also verified Swift behaviour [with a Swift interpreter](https://swiftfiddle.com/?c=H4sIAAAAAAAAA4WQMWvDMBCF%2F8rjJglqQ4dACZRs3bq06y2uIzsC%2BxSkc5MQ%2FN8rtVGGLEXLSffuu6d3pZ625OdjiIq3sMi%2BUx%2BE8zlGL2qYmqbBlxu9IFdMtvSGENEfuoj8yuTPaT4tTLiyAJNTRDe686dGvOa22TGborY2aYaOTEXnh5s0LZNmocbLDh9l0NRx2w4%2BJn3vtD8YL9sMmy%2BoEPu3D6hOb6hirm4sxZ1W3JRrUdnfnwAr3JTcI0kC%2FoPVeZb1MS0n%2B3tW9ETfOeFN%2B9I%2B0%2FoDsEAmXW8BAAA%3D), where `u` is not supported and it appears `w` is not yet supported either - but in the case of `w` it is at least clear what it should / will mean in a Swift regex.

Reply:
> The list comes from the [`NSRegularExpression`](https://developer.apple.com/documentation/foundation/nsregularexpression) doc (search for `(?ismwx-ismwx: ... )`), though I made an assumption that [`Regex`](https://developer.apple.com/documentation/swift/regex) works exactly the same way despite, as you noted, the doc not saying.

I think this might be good to note in a comment for future maintainers.

Reply:
Update: upon further investigation it appears the situation is more complicated, there are some more flags and differences between `Regex` and `NSRegularExpression`.  I'll add a documenting comment and expand the set of flags we accept.

Reply:
I've updated the code.